core/src/main/java/com/linecorp/armeria/common/stream/StreamWriterOutputStream.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.function.Function;
+
+import com.linecorp.armeria.common.HttpData;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+
+final class StreamWriterOutputStream<T> extends OutputStream {
+
+    private final StreamWriter<T> writer;
+    private final Function<? super HttpData, ? extends T> httpDataConverter;
+    private ByteBuf buffer;
+    private final int maxBufferSize;
+    private boolean closed;
+
+    StreamWriterOutputStream(StreamWriter<T> writer,
+                             Function<? super HttpData, ? extends T> httpDataConverter,
+                             int maxBufferSize) {
+        requireNonNull(writer, "writer");
+        requireNonNull(httpDataConverter, "httpDataConverter");
+        checkArgument(maxBufferSize > 0, "maxBufferSize should be positive");
+        this.writer = writer;
+        this.httpDataConverter = httpDataConverter;
+        this.maxBufferSize = maxBufferSize;
+        buffer = Unpooled.buffer(maxBufferSize, maxBufferSize);
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+        ensureOpen();
+        maybeDrain();
+        buffer.writeByte(b);
+    }
+
+    @Override
+    public void write(byte[] bytes, int off, int len) throws IOException {
+        requireNonNull(bytes, "bytes");
+        checkArgument(off >= 0, "offset should be non negative");
+        checkArgument(len >= 0, "length should be non negative");
```

Comment:
Trustin shared this philosophy on API design at https://www.youtube.com/watch?v=0eQbsVLxmMk.
You may want to check it out. 😉
```suggestion
        checkArgument(off >= 0, "off %s ( expected: >= 0)", off);
        checkArgument(len >= 0, "len %s ( expected: >= 0)", len);
```